Memory snapshots from C++

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86241
* __->__ #86190
* #86254

Sometimes the driving process want to save memory snapshots but isn't Python.
Add a simple API to turn it on without python stack traces. It still
saves to the same format for the vizualization and summary scripts, using
the C++ Pickler.

zdevito committed:
Memory snapshots from C++

Sometimes the driving process want to save memory snapshots but isn't Python.
Add a simple API to turn it on without python stack traces. It still
saves to the same format for the vizualization and summary scripts, using
the C++ Pickler.

[ghstack-poisoned]

zdevito committed:
Update on "Memory snapshots from C++"


Sometimes the driving process want to save memory snapshots but isn't Python.
Add a simple API to turn it on without python stack traces. It still
saves to the same format for the vizualization and summary scripts, using
the C++ Pickler.

[ghstack-poisoned]

zdevito committed:
Update on "Memory snapshots from C++"


Sometimes the driving process want to save memory snapshots but isn't Python.
Add a simple API to turn it on without python stack traces. It still
saves to the same format for the vizualization and summary scripts, using
the C++ Pickler.

[ghstack-poisoned]

ezyang commented on torch/csrc/jit/serialization/pickler.h:
```diff
@@ -274,6 +276,7 @@ class TORCH_API Pickler {
   std::unordered_map<std::string, uint32_t> memoized_globals_map_;
   std::unordered_map<std::string, uint32_t> memoized_strings_map_;
   std::unordered_map<std::string, uint32_t> memoized_devices_map_;
+  bool tag_aggregates_;
```
not sure what this field is doing

ezyang commented on test/test_cuda.py:
```diff
@@ -4694,6 +4694,34 @@ def test_raises_oom(self):
+        try:
+            m.record(True)
+            t = torch.rand(311, 411, device='cuda')
+            mem = pickle.loads(m.do_snapshot())
```
nice

ezyang commented on torch/csrc/cuda/memory_snapshot.cpp:
```diff
@@ -0,0 +1,125 @@
+      enabled ? blank_context : nullptr);
+}
+
+std::vector<char> _memory_snapshot_pickled() {
```
going into a vector<char> is kinda weird but OK

zdevito replied:
A std::string instead? I don't want it to be a file because these things might need to be written via API calls rather than directly to a file.

ezyang replied:
going to a file stream would be the normal thing, but I think this is fine, we can reevaluate if this ends up being too big to store in memory (it shouldn't ever, tbh)

ezyang commented on torch/csrc/cuda/memory_snapshot.cpp:
```diff
@@ -0,0 +1,125 @@
+    segmentDict.insert(blocks_s, blocks);
+
+    return segmentDict;
+  };
```
Wondering if we can dedupe this code with Python. We generate an IValue, and then convert it to Python. Well, it's not that long.

zdevito committed:
Update on "Memory snapshots from C++"


Sometimes the driving process want to save memory snapshots but isn't Python.
Add a simple API to turn it on without python stack traces. It still
saves to the same format for the vizualization and summary scripts, using
the C++ Pickler.

[ghstack-poisoned]

zdevito committed:
Update on "Memory snapshots from C++"


Sometimes the driving process want to save memory snapshots but isn't Python.
Add a simple API to turn it on without python stack traces. It still
saves to the same format for the vizualization and summary scripts, using
the C++ Pickler.

[ghstack-poisoned]

zdevito committed:
Update on "Memory snapshots from C++"


Sometimes the driving process want to save memory snapshots but isn't Python.
Add a simple API to turn it on without python stack traces. It still
saves to the same format for the vizualization and summary scripts, using
the C++ Pickler.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge
